### ISSUE TITLE
validator failing intermittently

### ISSUE THREAD

**Issue by machow** [cal-itp/data-infra#66]
May be related to an addressed gcsfs issue https://github.com/dask/gcsfs/issues/315. Going to bump gtfs-validator-api to use a newer gcsfs version. Otherwise, I'll investigate whether its a kubernetes resource issue. 

From looking at the airflow logs, it seems to fail when either using fs.get, or fs.pipe (depending on how far it gets).

**Comment by hunterowens** [comment]
ok cool. saw that you cut the new release

**Comment by hunterowens** [comment]
man it is wild how much less mature the GCSFS compared to s3fs / boto tooling is.

**Comment by machow** [comment]
yeah--for sure, big ups to dask for maintaining gcsfs. I'm still pretty curious what the best route forward is for pulling out the main bigquery ops we'll want to do from airflow is... (or whether we should just use their hooks outside of airflow...)

**Comment by machow** [comment]
Hmm.. no luck with upgrading gcsfs. It works when we run kube pods via the dev pipeline, and when I run locally. I'm going to check w/ @lottspot to see if we can pair on checking kube stuff next week. Even though dev and prod should be running in the same cluster, the task seems to run longer on prod (and then fail).

In the meantime, I'll..

* do a run on dev and prod that doesn't delete the pods, so we can take a closer look at them.
* Use docker on my laptop each morning, to run the validator over the prod bucket.

**Comment by machow** [comment]
Reviewed w/ @lottspot yesterday--and he pointed out that the airflow workers are using a high proportion of their memory. I've been taking a closer look at metrics, and wanted to dump some findings.

**tl;dr: we should either beef up our workers, or cut the number of concurrent celery workers (e.g. from 6 to 2). I don't know how to stop the growing memory usage. Once @lottspot is done with the GTFS-RT work, maybe we can dig in to this + the pricing story (#45) further.**

* our custom pod task (running the gtfs-validator) doesn't much memory (~200mb), BUT..
* there seems to be growing memory usage in worker pods, caused by a container called gcs-syncd.

Before I show graphs, here are some resources:

* [upgrading gke nodes](https://cloud.google.com/composer/docs/how-to/managing/updating#upgrading_the_machine_type_for_gke_nodes)

I'll break this into two sections:

* overview: lots of graphs on memory usage
* possible causes

## Overview

| graph |  |
| --- | --- |
| **custom pod task memory** | ![image](https://user-images.githubusercontent.com/2574498/115574957-18ccd700-a290-11eb-9a76-86b799fb112f.png) | 
| **airflow worker pod memory** |  ![image](https://user-images.githubusercontent.com/2574498/115575565-adcfd000-a290-11eb-9c21-5fd28aad240a.png) |
| **node memory over time** | ![image](https://user-images.githubusercontent.com/2574498/115577936-d3f66f80-a292-11eb-9ad6-6f113b7ae72c.png) |
| **kubectl top nodes** | ![image](https://user-images.githubusercontent.com/2574498/115577597-91cd2e00-a292-11eb-9d8a-53352abcc604.png) |
| **container memory usage** |  ![image](https://user-images.githubusercontent.com/2574498/115580191-de196d80-a294-11eb-8060-a111e322bff8.png) <br><br> two major consumers are containers in the worker pod. green is the worker, and purple is something called gcs-syncd. It looks like gcs-syncd is what increases memory (though workers still don't have a ton of memory). |

## Possible causes

Basically, it comes down to

* I have no idea how to stop the growing memory usage of gcs-synchd
* we can cut memory usage in the worker by lower the number of concurrent celery workers

**Logs (no?)**. At first I thought it might be because google syncs keeps a bucket with logs, dags, etc.. that it mounts in workers. The logs weren't too big (~200Mb). I deleted the larger logs, but didn't notice a change in gcs-syncd...

**Tmp files (probably not)**. Then, I remembered that the pipeline has always downloaded data into a tmp directory, but never deleted it. These accumulate over time, but should be on disk?

![image](https://user-images.githubusercontent.com/2574498/115583639-12425d80-a298-11eb-87a5-0f17d5872329.png)

**Celery workers (yes)**.  It looks like each of the 6 celery workers takes about 150 Mb of our ~2.5Gb. We can lower the number of concurrent workers (e.g. even to 2 would be fine I think), or beef up memory.

![image](https://user-images.githubusercontent.com/2574498/115587225-6c90ed80-a29b-11eb-8fbf-5c50adbe9d24.png)

**Comment by hunterowens** [comment]
I have nothing more than, huh, wild. Let's look into the worker option first?

**Comment by machow** [comment]
So far, so good--looks like workers are using ~600Mb less memory, and hopefully w/ only 2 concurrent tasks things won't go out of control (I think they'll have 1.5Gbs to split between tasks, which should be fine for what we're doing rn?).
